Record the replication timeline_id in postgres sources to detect restores from backup at the upstream

rjobanp opened the pull request:
<!--
Describe the contents of the PR briefly but completely.

If you write detailed commit messages, it is acceptable to copy/paste them
here, or write "see commit messages for details." If there is only one commit
in the PR, GitHub will have already added its commit message above.
-->

### Motivation

  * This PR fixes a recognized bug: Fixes https://github.com/MaterializeInc/materialize/issues/7869


<!--
Which of the following best describes the motivation behind this PR?


    [Ensure issue is linked somewhere.]

  * This PR adds a known-desirable feature.

    [Ensure issue is linked somewhere.]

  * This PR fixes a previously unreported bug.

    [Describe the bug in detail, as if you were filing a bug report.]

  * This PR adds a feature that has not yet been specified.

    [Write a brief specification for the feature, including justification
     for its inclusion in Materialize, as if you were writing the original
     feature specification.]

   * This PR refactors existing code.

    [Describe what was wrong with the existing code, if it is not obvious.]
-->

### Tips for reviewer

I've left a couple of TODOs for improvements that I plan to do as follow ups. Feel free to call these out if you think I should include them in this review:
- This will only apply for newly created sources until we backfill the timeline-id on existing ones
- This currently only sets each subsource into a stalled state; the error does not propagate up to the postgres source itself:
```
materialize=> SELECT * FROM mz_internal.mz_source_statuses;
 id  |            name            |   type    |   last_status_change_at    | status  |
                  error                                  |                                           d
etails
-----+----------------------------+-----------+----------------------------+---------+----------------
---------------------------------------------------------+--------------------------------------------
-------------------------------------------------
 u5  | t1                         | subsource | 2023-12-07 20:49:06.148+00 | running |
                                                         |
 u6  | t2                         | subsource | 2023-12-07 20:49:06.38+00  | running |
                                                         |
 u3  | ten                        | subsource | 2023-12-07 20:49:06.144+00 | running |
                                                         |
 u9  | mz_source_progress         | progress  |                            | created |
                                                         |
 u10 | mz_source                  | postgres  | 2023-12-07 20:49:06.144+00 | running |
                                                         |
 u7  | alter_fail_drop_col        | subsource | 2023-12-07 20:49:06.151+00 | running |
                                                         |
 u4  | t0                         | subsource | 2023-12-07 20:49:06.141+00 | stalled | postgres: The r
eplication timeline id has changed. Expected 1 but got 2 | {"namespaced":{"postgres":"The replication
timeline id has changed. Expected 1 but got 2"}}
 u8  | alter_fail_drop_constraint | subsource | 2023-12-07 20:49:06.141+00 | stalled | postgres: The r
eplication timeline id has changed. Expected 1 but got 2 | {"namespaced":{"postgres":"The replication
timeline id has changed. Expected 1 but got 2"}}
(8 rows)

```

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

sploiselle commented on src/postgres-util/src/replication.rs:
```diff
@@ -141,3 +142,37 @@ pub async fn drop_replication_slots(
+        &replication_client.simple_query("IDENTIFY_SYSTEM").await?[0]
+    {
+        match r.get("timeline").expect("Returns one row").parse::<i8>() {
+            Ok(timeline_id) => Ok(timeline_id as u32),
```
We don't allow `as` functions––you can look in `mz_ore` for these types of fallible conversions.

sploiselle commented on src/postgres-util/src/replication.rs:
```diff
@@ -141,3 +142,37 @@ pub async fn drop_replication_slots(
+///
+/// Will start a replication connection to allow using the IDENTIFY_SYSTEM call
+/// for fetching the timeline ID.
+pub async fn get_timeline_id(
```
This extra wrapping function seems extraneous to me––we only need `get_timeline_id_with_client`, though I'd change that function's name to this one

sploiselle commented on src/postgres-util/Cargo.toml:
```diff
@@ -35,8 +35,8 @@ workspace-hack = { version = "0.0.0", path = "../workspace-hack", optional = tru
 default = ["workspace-hack", "replication", "schemas", "tunnel"]
-replication = ["tunnel"]
 schemas = ["prost", "serde", "proptest", "mz-proto", "tunnel"]
+replication = ["tunnel"]
```
Did this get auto-sorted into this order?

rjobanp replied:
Nope, this was my mistake when resolving a merge conflict - switching back!

sploiselle commented on src/postgres-util/src/replication.rs:
```diff
@@ -141,3 +142,37 @@ pub async fn drop_replication_slots(
+pub async fn get_timeline_id_with_client(
+    replication_client: &Client,
+) -> Result<u32, PostgresError> {
+    if let SimpleQueryMessage::Row(r) =
```
This seems like we should always expect this to return a row.

sploiselle commented on src/storage-controller/src/lib.rs:
```diff
@@ -520,6 +520,9 @@ where
         // self.upsert_collection_metadata(&mut durable_metadata, remap_shard_migration_delta)
         //     .await;
+
+        // TODO: backfill the timeline-id for postgres sources that were created
```
The backfill will happen in a catalog migration, or you could try to do it reactively when someone uses `ALTER SOURCE`.

sploiselle replied:
I pushed some commits to take care of this but neglected to remove the TODO––will take care of that once CI finishes.

sploiselle replied:
Oh you'd already removed it--ty

sploiselle commented on src/storage-types/src/sources.rs:
```diff
@@ -2227,13 +2227,18 @@ impl RustType<ProtoPostgresSourceConnection> for PostgresSourceConnection {
         ProtoPostgresSourcePublicationDetails {
             tables: self.tables.iter().map(|t| t.into_proto()).collect(),
             slot: self.slot.clone(),
+            timeline_id: self.timeline_id.clone().unwrap_or_default(),
```
I don't think `unwrap_or_default` is the right call here––you can mark the proto field to be `optional`, in which case it will accept an option here.

sploiselle commented on src/storage-types/src/sources.rs:
```diff
@@ -2245,6 +2250,12 @@ impl RustType<ProtoPostgresSourcePublicationDetails> for PostgresSourcePublicati
                 .map(mz_postgres_util::desc::PostgresTableDesc::from_proto)
                 .collect::<Result<_, _>>()?,
             slot: proto.slot,
+            // The minimum value for timeline_id is 1, so we use 0 to indicate an unknown timeline
```
I would just leave this in terms of concrete types and avoid dealing with magical sentinels.

sploiselle commented on src/storage/src/source/postgres.rs:
```diff
@@ -286,6 +286,8 @@ pub enum DefiniteError {
     MissingColumn,
     #[error("failed to parse COPY protocol")]
     InvalidCopyInput,
+    #[error("The replication timeline id has changed. Expected {0} but got {1}")]
```
the "style" of the other messages here is a good pattern to follow, e.g. maybe something like "invalid timeline ID from PostgreSQL server"

sploiselle commented on src/storage/src/source/postgres.rs:
```diff
@@ -286,6 +286,8 @@ pub enum DefiniteError {
     #[error("failed to parse COPY protocol")]
     InvalidCopyInput,
+    #[error("The replication timeline id has changed. Expected {0} but got {1}")]
+    InvalidTimelineId(u32, u32),
```
Rust supports more structured errors where you can name the fields––given that it isn't obvious which is which to the reader, I'd suggest using that e.g.

```rust
Enum {
  expected: u32,
  actual: u32,
}
```

sploiselle commented on src/storage/src/source/postgres/replication.rs:
```diff
@@ -260,6 +261,9 @@ pub(crate) fn render<G: Scope<Timestamp = MzOffset>>(
                 Err(err) => {
                     // If the replication stream cannot be obtained in a definite way there is
                     // nothing else to do. These errors are not retractable.
+                    // TODO: Some definite errors from the stream should put the entire source
```
We don't want to do this because we don't want to push data into the primary source's data shard––ideally, we should just make the shard totally empty.

I think the right approach is to start returning definite errors in the health stream and ensure we don't overwrite definite errors.

sploiselle commented on src/storage/src/source/postgres/replication.rs:
```diff
@@ -390,6 +395,16 @@ async fn raw_stream<'a>(
 > {
+    // Skip the timeline ID check for sources without a known timeline ID
+    // (sources created before the timeline ID was added to the source details)
+    if let Some(expected_timeline_id) = timeline_id {
```
I think this check wants to look more like where we ensure the replication slot––I have some changes in flight for that in #23746 as inspiration (still fighting with the change)

rjobanp replied:
I'm not sure I'm following -- did you mean the check to ensure the replication slot or the publication?

The `ensure_replication_slot` check looks like it returns a TransientError which I don't think makes sense in this context. In your PR it looks like the equivalent of the `ensure_publication_exists` check would involve moving this timeline check to `render` and to do:
```
            if let Err(err) = super::ensure_replication_timeline_id(&replication_client, &connection.publication_details.timeline_id).await? {
                // If the timeline changes there is nothing else to do. These errors
                // are not retractable.
                for &oid in table_info.keys() {
                    let update = ((oid, Err(err.clone())), MzOffset::from(u64::MAX), 1);
                    data_output.give(&data_cap_set[0], update).await;
                }
                return Ok(());
            }
```
but this is effectively the same as what it's doing now since the DefiniteError returned from the `raw_stream` wrapped result is already being handled the same way:
https://github.com/MaterializeInc/materialize/blob/bbe1b090e55fd53f3ebb4d75aa3e8953021e57dd/src/storage/src/source/postgres/replication.rs#L258-L270

sploiselle replied:
Sorry I meant publication! And you're totally right––wondering now why that check exists outside of the function? Seems like it would be cleaner (and not have caused the confusion in my comment) if it were inside.

sploiselle commented on test/pg-cdc-resumption/mzcompose.py:
```diff
@@ -216,3 +233,57 @@ def pg_out_of_disk_space(c: Composition) -> None:
     c.run("testdrive", "delete-rows-t2.td", "alter-table.td", "alter-mz.td")
+
+
+def backup_restore_pg(c: Composition) -> None:
```
Whoa this looks very daunting! Great work getting this test in here.

petrosagg reviewed (commented):
This is in great shape! A few minor details to iron and it should be good to go

nrainer-materialize commented:
[Nightly build](https://buildkite.com/materialize/nightlies/builds/5548) is green 👍

rjobanp commented:
Thanks for all the prompt reviews - I've addressed most of the feedback and resolved those conversations; left open the ones that might need more input

sploiselle commented:
@rjobanp Just pushed a few commits to this branch to handle the timeline ID backfill. Didn't remove your existing TODOs, but will do that if this doesn't pass CI.

sploiselle reviewed (approved):
Great work––really appreciate the effort and elbow grease that went into ensuring this got thoroughly tested. This has been a real blind spot of ours for a while and I'm relieved to see this land.

rjobanp commented:
@sploiselle thanks for pushing the commits to add the backfill! Any ideas on a good way to test that? I'm assuming I can't easily write a test case that runs MZ at version <0.80, creates a source, and then again at 0.80 to confirm the timeline-id is backfilled...

sploiselle commented:
We can definitely do the process you mention (or the equivalent of it) using the `upgrade-matrix` testing framework. Have you checked out that framework yet? If not, glad to go over it with you. What I'd normally recommend doing is adding a line test to the `validate` section of the check (e.g. `misc/python/materialize/checks/all_checks/pg_cdc.py`) that has different behaviors depending on the version. For example, we could ensure that old versions don't have the timeline ID, but new versions do. When this check runs in the upgrade matrix, we would run it where a source is created in some version < 80, and then where it's brought up in version 80––ensuring that the migration took place.

However, the challenge here is that we store the timeline ID in the `DETAILS` option of the PG source...which gets persisted as a hex-encoded byte string of the protobuf encoded struct. So not really a meaingful way to check that it's written to the catalog.

FWIW, if this was a very important piece of data, there are more involved and exotic solutions to the problem––e.g. we could create a new system table called `mz_pg_source_publication_details` that we could write the data like the timeline ID to. That seems overkill in this situation, but thought I'd mention it just as an illustration of how else we might be able to tackle this.

rjobanp commented:
> For example, we could ensure that old versions don't have the timeline ID, but new versions do. When this check runs in the upgrade matrix, we would run it where a source is created in some version < 80, and then where it's brought up in version 80––ensuring that the migration took place.

Awesome, the upgrade-matrix framework seems perfect for this! I suppose I could just validate that that `DETAILS` value in the `CREATE SOURCE` statement is different between versions, and not decode it? That would validate that the migration performed some update and didn't fail, which seems like it should be sufficient

sploiselle commented:
The checks themselves are stateless, so there's no way to remember the source's last-known `DETAILS` length.
